File: php/src/ParserConfiguration.php
<?php

namespace LogicAndTrick\WikiCodeParser;

use LogicAndTrick\WikiCodeParser\Elements\MdCodeElement;
use LogicAndTrick\WikiCodeParser\Elements\MdColumnsElement;
use LogicAndTrick\WikiCodeParser\Elements\MdHeadingElement;
use LogicAndTrick\WikiCodeParser\Elements\MdLineElement;
use LogicAndTrick\WikiCodeParser\Elements\MdListElement;
use LogicAndTrick\WikiCodeParser\Elements\MdPanelElement;
use LogicAndTrick\WikiCodeParser\Elements\MdQuoteElement;
use LogicAndTrick\WikiCodeParser\Elements\MdTableElement;
use LogicAndTrick\WikiCodeParser\Elements\PreElement;
use LogicAndTrick\WikiCodeParser\Elements\RefElement;
use LogicAndTrick\WikiCodeParser\Processors\AutoLinkingProcessor;
use LogicAndTrick\WikiCodeParser\Processors\MarkdownTextProcessor;
use LogicAndTrick\WikiCodeParser\Processors\NewLineProcessor;
use LogicAndTrick\WikiCodeParser\Processors\SmiliesProcessor;
use LogicAndTrick\WikiCodeParser\Processors\TrimWhitespaceAroundBlockNodesProcessor;
use LogicAndTrick\WikiCodeParser\Tags\CodeTag;
use LogicAndTrick\WikiCodeParser\Tags\FontTag;
use LogicAndTrick\WikiCodeParser\Tags\ImageTag;
use LogicAndTrick\WikiCodeParser\Tags\LinkTag;
use LogicAndTrick\WikiCodeParser\Tags\PreTag;
use LogicAndTrick\WikiCodeParser\Tags\QuickLinkTag;
use LogicAndTrick\WikiCodeParser\Tags\QuoteTag;
use LogicAndTrick\WikiCodeParser\Tags\SpoilerTag;
use LogicAndTrick\WikiCodeParser\Tags\Tag;
use LogicAndTrick\WikiCodeParser\Tags\VaultEmbedTag;
use LogicAndTrick\WikiCodeParser\Tags\WikiArchiveTag;
use LogicAndTrick\WikiCodeParser\Tags\WikiBookTag;
use LogicAndTrick\WikiCodeParser\Tags\WikiCategoryTag;
use LogicAndTrick\WikiCodeParser\Tags\WikiCreditTag;
use LogicAndTrick\WikiCodeParser\Tags\WikiFileTag;
use LogicAndTrick\WikiCodeParser\Tags\WikiImageTag;
use LogicAndTrick\WikiCodeParser\Tags\WikiLinkTag;
use LogicAndTrick\WikiCodeParser\Tags\WikiYoutubeTag;
use LogicAndTrick\WikiCodeParser\Tags\YoutubeTag;

class ParserConfiguration
{
    public static function Twhl() : ParserConfiguration {
        $conf = new ParserConfiguration();

        // Standard inline
        $conf->tags[] = (new Tag('b', 'strong'))->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new Tag('i', 'em'))->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new Tag('u', 'span', 'underline'))->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new Tag('s', 'span', 'strikethrough'))->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new Tag('green', 'span', 'green'))->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new Tag('blue', 'span', 'blue'))->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new Tag('red', 'span', 'red'))->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new Tag('purple', 'span', 'purple'))->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new Tag('yellow', 'span', 'yellow'))->WithScopes('inline', 'excerpt');

        // Standard block
        $conf->tags[] = new PreTag();
        $conf->tags[] = (new Tag('h', 'h3'))->WithBlock(true);


        // Links
        $conf->tags[] = (new LinkTag())->WithScopes('excerpt');
        $conf->tags[] = (new LinkTag())->WithScopes('excerpt')->WithToken('email');
        $conf->tags[] = new QuickLinkTag();
        $conf->tags[] = new WikiLinkTag();
        $conf->tags[] = new WikiFileTag();

        // Embedded
        $conf->tags[] = new ImageTag();
        $conf->tags[] = (new ImageTag())->WithToken('simg')->WithBlock(false);
        $conf->tags[] = new WikiImageTag();
        $conf->tags[] = new YoutubeTag();
        $conf->tags[] = new WikiYoutubeTag();
        $conf->tags[] = new VaultEmbedTag();

        // Custom
        $conf->tags[] = new QuoteTag();
        $conf->tags[] = (new FontTag())->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new WikiCategoryTag())->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new WikiBookTag())->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new WikiCreditTag())->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new WikiArchiveTag())->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new SpoilerTag())->WithScopes('inline', 'excerpt');
        $conf->tags[] = (new CodeTag())->WithScopes('excerpt');

        // Elements
        $conf->elements[] = new MdCodeElement();
        $conf->elements[] = new PreElement();
        $conf->elements[] = new MdHeadingElement();
        $conf->elements[] = new MdLineElement();
        $conf->elements[] = new MdQuoteElement();
        $conf->elements[] = new MdListElement();
        $conf->elements[] = new MdTableElement();
        $conf->elements[] = new MdPanelElement();
        $conf->elements[] = new MdColumnsElement();
        $conf->elements[] = new RefElement();

        // Processors
        $conf->processors[] = new MarkdownTextProcessor();
        $conf->processors[] = new AutoLinkingProcessor();
        $conf->processors[] = (new SmiliesProcessor())->AddDefault();
        $conf->processors[] = new TrimWhitespaceAroundBlockNodesProcessor();
        $conf->processors[] = new NewLineProcessor();

        return $conf;
    }

    public array $elements;
    public array $tags;
    public array $processors;

    public function __construct() {
        $this->elements = [];
        $this->tags = [];
        $this->processors = [];
    }
}

File: csharp/LogicAndTrick.WikiCodeParser/ParserConfiguration.cs
using System.Collections.Generic;
using LogicAndTrick.WikiCodeParser.Elements;
using LogicAndTrick.WikiCodeParser.Processors;
using LogicAndTrick.WikiCodeParser.Tags;

namespace LogicAndTrick.WikiCodeParser
{
    public class ParserConfiguration
    {
        /// <summary>
        /// Create an instance of configuration that matches the configuration of the TWHL website.
        /// </summary>
        /// <returns>Default configuration</returns>
        public static ParserConfiguration Twhl()
        {
            var conf = new ParserConfiguration();

            // Standard inline
            conf.Tags.Add(new Tag("b", "strong").WithScopes("inline", "excerpt"));
            conf.Tags.Add(new Tag("i", "em").WithScopes("inline", "excerpt"));
            conf.Tags.Add(new Tag("u", "span", "underline").WithScopes("inline", "excerpt"));
            conf.Tags.Add(new Tag("s", "span", "strikethrough").WithScopes("inline", "excerpt"));
            conf.Tags.Add(new Tag("green", "span", "green").WithScopes("inline", "excerpt"));
            conf.Tags.Add(new Tag("blue", "span", "blue").WithScopes("inline", "excerpt"));
            conf.Tags.Add(new Tag("red", "span", "red").WithScopes("inline", "excerpt"));
            conf.Tags.Add(new Tag("purple", "span", "purple").WithScopes("inline", "excerpt"));
            conf.Tags.Add(new Tag("yellow", "span", "yellow").WithScopes("inline", "excerpt"));

            // Standard block
            conf.Tags.Add(new PreTag());
            conf.Tags.Add(new Tag("h", "h3").WithBlock(true));

            // Links
            conf.Tags.Add(new LinkTag().WithScopes("excerpt"));
            conf.Tags.Add(new LinkTag().WithScopes("excerpt").WithToken("email"));
            conf.Tags.Add(new QuickLinkTag());
            conf.Tags.Add(new WikiLinkTag());
            conf.Tags.Add(new WikiFileTag());

            // Embedded
            conf.Tags.Add(new ImageTag());
            conf.Tags.Add(new ImageTag().WithToken("simg").WithBlock(false));
            conf.Tags.Add(new WikiImageTag());
            conf.Tags.Add(new YoutubeTag());
            conf.Tags.Add(new WikiYoutubeTag());
            conf.Tags.Add(new VaultEmbedTag());

            // Custom
            conf.Tags.Add(new QuoteTag());
            conf.Tags.Add(new FontTag().WithScopes("inline", "excerpt"));
            conf.Tags.Add(new WikiCategoryTag().WithScopes("inline", "excerpt"));
            conf.Tags.Add(new WikiBookTag().WithScopes("inline", "excerpt"));
            conf.Tags.Add(new WikiCreditTag().WithScopes("inline", "excerpt"));
            conf.Tags.Add(new WikiArchiveTag().WithScopes("inline", "excerpt"));
            conf.Tags.Add(new SpoilerTag().WithScopes("inline", "excerpt"));
            conf.Tags.Add(new CodeTag().WithScopes("excerpt"));

            // Elements
            conf.Elements.Add(new MdCodeElement());
            conf.Elements.Add(new PreElement());
            conf.Elements.Add(new MdHeadingElement());
            conf.Elements.Add(new MdLineElement());
            conf.Elements.Add(new MdQuoteElement());
            conf.Elements.Add(new MdListElement());
            conf.Elements.Add(new MdTableElement());
            conf.Elements.Add(new MdPanelElement());
            conf.Elements.Add(new MdColumnsElement());
            conf.Elements.Add(new RefElement());
            
            // Processors
            conf.Processors.Add(new MarkdownTextProcessor());
            conf.Processors.Add(new AutoLinkingProcessor());
            conf.Processors.Add(new SmiliesProcessor().AddTwhl());
            conf.Processors.Add(new TrimWhitespaceAroundBlockNodesProcessor());
            conf.Processors.Add(new NewLineProcessor());

            return conf;
        }

        /// <summary>
        /// Create an instance of configuration that matches the configuration of the Snarkpit website.
        /// </summary>
        /// <returns>Default configuration</returns>
        public static ParserConfiguration Snarkpit()
        {
            var conf = new ParserConfiguration();

            // Standard inline
            conf.Tags.Add(new Tag("b", "strong").WithScopes("inline", "excerpt"));
            conf.Tags.Add(new Tag("i", "em").WithScopes("inline", "excerpt"));
            conf.Tags.Add(new Tag("u", "span", "underline").WithScopes("inline", "excerpt"));
            conf.Tags.Add(new Tag("s", "span", "strikethrough").WithScopes("inline", "excerpt"));

            // Standard block
            conf.Tags.Add(new PreTag());
            conf.Tags.Add(new Tag("center", "div", "text-center").WithBlock(true));
            // aligntag
            // listtag
            
            // Links
            conf.Tags.Add(new LinkTag().WithScopes("excerpt"));
            conf.Tags.Add(new LinkTag().WithScopes("excerpt").WithToken("email"));
            conf.Tags.Add(new QuickLinkTag());

            // Embedded
            conf.Tags.Add(new ImageTag());
            conf.Tags.Add(new ImageTag().WithToken("simg").WithBlock(false));
            conf.Tags.Add(new WikiImageTag());

            conf.Tags.Add(new YoutubeTag());
            conf.Tags.Add(new WikiYoutubeTag());

            // Custom
            conf.Tags.Add(new QuoteTag());
            // color
            // size
            conf.Tags.Add(new SpoilerTag().WithScopes("inline", "excerpt"));

            // Elements
            conf.Elements.Add(new MdCodeElement());
            conf.Elements.Add(new PreElement { Token = "code" });
            conf.Elements.Add(new MdHeadingElement());
            conf.Elements.Add(new MdLineElement());
            conf.Elements.Add(new MdQuoteElement());
            conf.Elements.Add(new MdListElement());
            conf.Elements.Add(new MdTableElement());
            conf.Elements.Add(new MdPanelElement());
            conf.Elements.Add(new MdColumnsElement());
            conf.Elements.Add(new RefElement());

            // Processors
            conf.Processors.Add(new MarkdownTextProcessor());
            conf.Processors.Add(new AutoLinkingProcessor());
            conf.Processors.Add(new SmiliesProcessor().AddSnarkpit());
            conf.Processors.Add(new TrimWhitespaceAroundBlockNodesProcessor());
            conf.Processors.Add(new NewLineProcessor());

            return conf;
        }

        public List<Element> Elements { get; }
        public List<Tag> Tags { get; }
        public List<INodeProcessor> Processors { get; }

        public ParserConfiguration()
        {
            Elements = new List<Element>();
            Tags = new List<Tag>();
            Processors = new List<INodeProcessor>();
        }
    }
}
